### ISSUE TITLE
Derivative for _ctc_loss_backward

### ISSUE THREAD

**Issue by grazder** [pytorch/pytorch#70108]
### 🚀 The feature, motivation and pitch

Discussion motivation: https://discuss.pytorch.org/t/higher-order-gradients-of-ctcloss/35019

Hi! I'm trying to do double backward for CTC loss function and I can't do it due to `Runtime Error: derivative for _ctc_loss_backward is not implemented` error.

I'm implementing https://arxiv.org/abs/2102.08098 paper for a network with CTC loss. So I can't do it in a neutral way with default `torch.nn.functional.ctc_loss`.

### Alternatives

As an alternatives, I'm considering using different losses or hand-written ctc loss realization which are not optimal for our network, so it can cause not optimal optimization process.

**Comment by albanD** [comment]
Hi,

Thanks you for the feature request. We would definitely be happy to accept a PR adding this!

**Comment by vadimkantorov** [comment]
@albanD does functorch by chance have a finite difference approximation? e.g. it may be nice to have a standardized finite differences for double backward (maybe already implemented in gradgradcheck?)

**Comment by albanD** [comment]
gradcheck/gradgradcheck can be used to check this yes.
functorch is based on the formulas in pytorch so we cannot really use it for checking.

**Comment by vadimkantorov** [comment]
> we cannot really use it for checking.

I was talking not about checking per se, but about surfacing out the finite difference code that is done in gradgradcheck as public API.

If I understand well, it finite differences may be used for some ops that lack manual double backward

**Comment by vadimkantorov** [comment]
I agree, but for some scenarios it may still be useful - maybe even to validate other dependent code before writing a custom double backward.

**Comment by ZR-HH** [comment]
I also meet the same error, when I used ctc_loss for meta learning algorithm, the error appeared. But if I used the kiv_loss, there was not anything wrong. If you have some suggest for my questions? Thanks

**Comment by Stanwang1210** [comment]
Have anyone solve this issues ?

**Comment by vadimkantorov** [comment]
One can reformulate CTC as first computing the optimal alignment path and then calculating cross entropy, as in https://github.com/vadimkantorov/ctc

Then it should be able to do double-backward through that cross-entropy calculation 

But I'm not sure if this is the correct gradient for double-backward in general

**Comment by Stanwang1210** [comment]
@vadimkantorov 
Oh! That's amazing! I really appreciate that!

But is there any way that we can easily integrate this code with the original CTC?
I think maybe we can directly replace the F.ctc_loss in nn.CTC_Loss like you did in the [example](https://github.com/vadimkantorov/ctc/blob/master/example.py) ?
Can you provide some suggestion to use it?

**Comment by vadimkantorov** [comment]
You can find in this example the cross-entropy expression using the calculated targets and double-backward through that (keeping the targets constant). But again, I'm not sure if this approximation of the double-gradient is accurate

**Comment by Stanwang1210** [comment]
@vadimkantorov 
Thanks for you help again!

I fix the in-place operation in ctc_loss myself with the [suggestion](https://github.com/vadimkantorov/ctc/issues/2) from @Tzeviya
However, I'm wondering whether you have solve the alignment error in your issues?
If yes, please let me know

**Comment by vadimkantorov** [comment]
@Stanwang1210 yep, fixed the alignment issue in my repo. regarding your in-place fixes, have you tried to do double-backward? (the vanilla `example.py` version in my repo runs fine now)
